modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,385 @@
+use crate::state::{
+    ChallengeEntry, CheckIn, Friend, Penalty, Vote, CHALLENGE_FRIENDS, CHALLENGE_LIST,
+    DAILY_CHECK_INS, NEXT_ID, VOTES,
+};
+
+pub fn execute_handler(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    msg: ChallengeExecuteMsg,
+) -> AppResult {
+    match msg {
+        ChallengeExecuteMsg::CreateChallenge { challenge } => {
+            create_challenge(deps, env, info, app, challenge)
+        }
+        ChallengeExecuteMsg::UpdateChallenge {
+            challenge_id,
+            challenge,
+        } => update_challenge(deps, env, info, app, challenge_id, challenge),
+        ChallengeExecuteMsg::CancelChallenge { challenge_id } => {
+            cancel_challenge(deps, info, app, challenge_id)
+        }
+        ChallengeExecuteMsg::AddFriendForChallenge {
+            challenge_id,
+            friend_name,
+            friend_address,
+        } => add_friend_for_challenge(
+            deps,
+            info,
+            &app,
+            challenge_id,
+            &friend_name,
+            &friend_address,
+        ),
+        ChallengeExecuteMsg::RemoveFriendForChallenge {
+            challenge_id,
+            friend_address,
+        } => remove_friend_from_challenge(deps, info, &app, challenge_id, friend_address),
+        ChallengeExecuteMsg::AddFriendsForChallenge {
+            challenge_id,
+            friends,
+        } => add_friends_for_challenge(deps, info, &app, challenge_id, friends),
+        ChallengeExecuteMsg::DailyCheckIn {
+            challenge_id,
+            metadata,
+        } => daily_check_in(deps, env, info, &app, challenge_id, metadata),
+        ChallengeExecuteMsg::CastVote { vote, challenge_id } => {
+            cast_vote(deps, env, info, &app, vote, &challenge_id)
+        }
+        ChallengeExecuteMsg::CountVotes { challenge_id } => {
+            count_votes(deps, info, env, &app, challenge_id)
+        }
+    }
+}
+
+/// Create new challenge
+fn create_challenge(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    challenge: ChallengeEntry,
+) -> AppResult {
+    // Only the admin should be able to create a challenge.
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    // Generate the challenge id
+    let id = NEXT_ID.update(deps.storage, |id| AppResult::Ok(id + 1))?;
+    let challenge_id = format!("challenge_{id}");
+
+    CHALLENGE_LIST.save(deps.storage, challenge_id.clone(), &challenge)?;
+
+    Ok(app.tag_response(
+        Response::new().add_attribute("challenge_id", challenge_id),
+        "create_challenge",
+    ))
+}
+
+/// Update an existing challenge  
+fn update_challenge(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    app: ChallengeApp,
+    challenge_id: String,
+    challenge: ChallengeEntry,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    // will return an error if the challenge doesn't exist
```

Comment:
The errors returned by CosmWasm are notorious for being uninformative so often it's better to parse them into something more human-readable.

For updating values in storage we often opt for an `Option` based struct. I.e. a `ChallengeEntryUpdate` struct of the like that has all its field be `Option`. That way when updating a value you don't have to re-affirm the values you dont want to change.

Reply:
> The errors returned by CosmWasm are notorious for being uninformative so often it's better to parse them into something more human-readable.

I guess the usual approach would be to use `.may_load().ok_or()?`. Because we want to have different error in case storage is screwed somehow

Reply:
@Buckram123 using `ok_or()?` was causing a compiler error because the returned error type does not directly convert to `AppError`, so I used `map_err()?` instead.